File: integrations/langgraph/typescript/src/middlewares/state-streaming.ts
/**
 * Custom middleware helpers for CopilotKit agents.
 */

import { createMiddleware } from "langchain";
import { BaseMessage, ToolMessage } from "@langchain/core/messages";

export interface StateItem {
  stateKey: string;
  tool: string;
  toolArgument: string;
}

/** Identity helper — exists purely for IDE type inference on the object literal. */
export const stateItem = (item: StateItem): StateItem => item;

/**
 * Middleware that injects `predict_state` metadata into model invocations so
 * that every `on_chat_model_stream` event carries it.
 *
 * Approach: wrap `request.model` with `model.withConfig({ metadata: {
 * predict_state } })` before passing it to the base handler. When the base
 * handler subsequently calls `bindTools()` on this RunnableBinding,
 * `_simpleBindTools` detects the RunnableBinding wrapper and creates a new
 * RunnableBinding that **preserves our config**. `RunnableBinding.invoke()`
 * then uses `mergeConfigs()` (which deep-merges metadata) to combine our
 * bound config with the LangGraph execution config, so `predict_state`
 * survives into every streaming event.
 */
export const stateStreamingMiddleware = (...items: StateItem[]) => {
  const predictState = items.map((i) => ({
    state_key: i.stateKey,
    tool: i.tool,
    tool_argument: i.toolArgument,
  }));

  /**
   * Return true if this model call may generate the initial tool call.
   * When the last message is a ToolMessage the tool has already run and
   * the model is being called for a follow-up response. Injecting
   * predict_state in that case would re-trigger streaming if the model
   * decides to call the same tool again, producing a duplicate stream.
   */
  const isPreToolCall = (request: { messages?: BaseMessage[] }): boolean => {
    const msgs = request?.messages ?? [];
    if (msgs.length === 0) return true;
    return !(msgs[msgs.length - 1] instanceof ToolMessage);
  };

  return createMiddleware({
    name: "StateStreamingMiddleware",
    wrapModelCall: async (request, handler) => {
      if (!isPreToolCall(request)) {
        return handler(request);
      }
      const modelWithState = request.model.withConfig({
        metadata: { predict_state: predictState },
      });
      return handler({ ...request, model: modelWithState });
    },
  });
};

File: integrations/langgraph/python/ag_ui_langgraph/middlewares/state_streaming.py
"""
Custom middleware helpers for ag-ui LangGraph agents.
"""
from dataclasses import dataclass
from typing import Any, Awaitable, Callable

try:
    from langchain.agents.middleware import AgentMiddleware, ModelRequest
    _MIDDLEWARE_AVAILABLE = True
except ImportError:
    AgentMiddleware = object  # type: ignore[assignment,misc]
    ModelRequest = None  # type: ignore[assignment,misc]
    _MIDDLEWARE_AVAILABLE = False

from langchain_core.messages import ToolMessage
from langchain_core.runnables.config import ensure_config, var_child_runnable_config


def _with_intermediate_state(config: dict, emit_intermediate_state: list) -> dict:
    metadata = {**config.get("metadata", {}), "predict_state": emit_intermediate_state}
    return {**config, "metadata": metadata}


@dataclass(frozen=True)
class StateItem:
    state_key: str
    tool: str
    tool_argument: str

class StateStreamingMiddleware(AgentMiddleware):
    def __init__(self, *items: StateItem) -> None:
        if not _MIDDLEWARE_AVAILABLE:
            raise ImportError(
                "StateStreamingMiddleware requires langchain>=1.2.0. "
                "Please upgrade: pip install 'langchain>=1.2.0'"
            )
        self._emit_intermediate_state = [
            {"state_key": i.state_key, "tool": i.tool, "tool_argument": i.tool_argument}
            for i in items
        ]

    def _is_pre_tool_call(self, request: ModelRequest) -> bool:
        """Return True if this model call precedes a tool call for the current turn.

        When the last message is a ToolMessage the tool has already run and the
        model is being called for a follow-up response.  Injecting
        emit_intermediate_state in that case causes predict_state streaming to
        fire again if the model decides to call the same tool a second time,
        producing a duplicate stream.
        """
        msgs = request.messages
        return not (msgs and isinstance(msgs[-1], ToolMessage))

    def wrap_model_call(
        self,
        request: ModelRequest,
        handler: Callable[[ModelRequest], Any],
    ) -> Any:
        if not self._is_pre_tool_call(request):
            return handler(request)
        config = _with_intermediate_state(ensure_config(), self._emit_intermediate_state)
        token = var_child_runnable_config.set(config)
        try:
            return handler(request)
        finally:
            var_child_runnable_config.reset(token)

    async def awrap_model_call(
        self,
        request: ModelRequest,
        handler: Callable[[ModelRequest], Awaitable[Any]],
    ) -> Any:
        pre = self._is_pre_tool_call(request)
        if not pre:
            return await handler(request)
        config = _with_intermediate_state(ensure_config(), self._emit_intermediate_state)
        token = var_child_runnable_config.set(config)
        try:
            return await handler(request)
        finally:
            var_child_runnable_config.reset(token)
